### DISCUSSION TITLE
Litestar consumes high amount of CPU

### DISCUSSION THREAD

**Discussion by 0asys** [litestar-org/litestar#2504] [Q&A] [answered]
### Description

Running litestar `litestar app run` with uvicorn consumes in an idling state on my machines 40-45% CPU.

Is this normal or is there a way to reduce the idling CPU consumption?

Attached a picture of cProfile. `python -m cProfile -o litestar.pstats ./.venv/bin/litestar run`

![litestar_profile](https://github.com/litestar-org/litestar/assets/1131885/ab23a525-f85e-479e-8a5d-981fe68c46e3)

### MCVE

```python
UVICORN_LOOP=uvloop litestar run                            
Using Litestar app from env: 'app.asgi:create_app'
Starting server process ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
┌──────────────────────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Litestar version             │ 2.2.1                                                                                                        │
│ Debug mode                   │ Disabled                                                                                                     │
│ Python Debugger on exception │ Disabled                                                                                                     │
│ CORS                         │ Enabled                                                                                                      │
│ CSRF                         │ Disabled                                                                                                     │
│ OpenAPI                      │ Enabled path=/schema                                                                                         │
│ Compression                  │ Disabled                                                                                                     │
│ Template engine              │ ViteTemplateEngine                                                                                           │
│ Static files                 │ path=/static dirs=.../src/app/domain/web/public html_mode=Enabled │
│ Middlewares                  │ JWTCookieAuthenticationMiddleware, SentryLitestarASGIMiddleware, middleware_factory                          │
└──────────────────────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
2023-10-23T19:12:38.780745Z [info     ] Started server process [4257]  [uvicorn.error] 
2023-10-23T19:12:38.781923Z [info     ] Waiting for application startup. [uvicorn.error] 
2023-10-23T19:12:38.789916Z [info     ] Application startup complete.  [uvicorn.error] 
2023-10-23T19:12:38.792650Z [info     ] Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit) [uvicorn.error] 
^C2023-10-23T19:15:04.770399Z [info     ] Shutting down                  [uvicorn.error] 
^C2023-10-23T19:15:04.910332Z [info     ] Finished server process [4257] [uvicorn.error] 
^C2023-10-23T19:15:04.974529Z [info     ] ASGI 'lifespan' protocol appears unsupported. [uvicorn.error] 
```
```


### Steps to reproduce

```bash
- Run `litestar run`
- Look process ressources
```

### Screenshots

![litestar_profile](https://github.com/litestar-org/litestar/assets/1131885/ab23a525-f85e-479e-8a5d-981fe68c46e3)

### Litestar Version

2.2.1
python3.11

### Platform

- [ ] Linux
- [X] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh Litestar dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2503">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2503/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/2503/pledge.svg">
</picture>
</a>

**Comment by provinzkraut** [comment]
I cannot reproduce this with a basic application. Can you provide a reproducible example? It’s really hard to debug this without actually know what your app looks like, but I’m fairly certain that there’s some kind of busy loop involved you’re not aware of.

**Reply by 0asys** [reply]
I can put togheter an MCVE.

in the meantime i found the culprit in the channels implementation:
https://github.com/litestar-org/litestar/blob/0d4fe01a9d361e09748f71aa7adf70672b97c822/litestar/channels/backends/memory.py#L69

this looping and waiting generates the CPU load.
If I disable the channels plugin or use the `RedisChannelBackend` with an subscription idle time of 1000 it is at the same level as running litestar without channels.

**Reply by provinzkraut** [reply, marked as answer]
@bpereto Thanks for the insightful report! I put up a fix over here #2547. Can you confirm this fixes your issue?
